Move logic for generated images from CSSValue types to the StyleImage types

weinig opened the pull request:
#### 32d57af5e2f83d93188e8ef0afcfa065bb35a41f
<pre>
Move logic for generated images from CSSValue types to the StyleImage types
<a href="https://bugs.webkit.org/show_bug.cgi?id=246594">https://bugs.webkit.org/show_bug.cgi?id=246594</a>
rdar://101225901

Reviewed by Darin Adler.

In a step toward creating a comprehensive layering between the CSSValue types
and the Style types they resolve to, this change moves all of the logic that
used to be in the CSSValue derived types for generated images into their new
Style counterparts. This leaves the CSSValue side as immutable representations
of the parsed CSS (or an immutable representation of a computed CSS property)
that is capable of only a few operations:
   - equality
   - serialization
   - resolution to Style counterpart

This initial move does a pretty close approximation of just moving the logic
and state, in some cases having the Style type continue to store CSSValues,
but subsequent changes should change this so that we do conversion to lower
level types at creation time (e.g. use Length rather than CSSPrimitiveValue).
The one major exception is that we now reference other images as StyleImages
rather than CSSValues in the places where the types require input images
such as StyleCrossfadeImage and StyleFilterImage. This puts us in a strong
position to fix a bunch of long standing issues with animating between
various combinations of nested images.

Due to the simplification on the CSSValue side, we no longer need the
CSSImageGeneratorValue intermediate class, so it has been removed.

This change does not update CSSImageValue in the same way as the generated
image classes, as that has additional complications that should be handled
on their own.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSImageGeneratorValue.cpp: Removed.
* Source/WebCore/css/CSSImageGeneratorValue.h: Removed.
Remove CSSImageGeneratorValue.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFilter):
(WebCore::crossfadeBlend):
(WebCore::blendFunc):
Update to use StyleImages (and subclasses thereof) rather than the CSSValue
types, making the logic much closer to other blend primitives and allowing
the removal of a lot of unnecessary temporary allocations.

* Source/WebCore/css/CSSCanvasValue.cpp:
* Source/WebCore/css/CSSCanvasValue.h:
* Source/WebCore/css/CSSCrossfadeValue.cpp:
* Source/WebCore/css/CSSCrossfadeValue.h:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFilterImageValue.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/CSSImageSetValue.cpp:
* Source/WebCore/css/CSSImageSetValue.h:
* Source/WebCore/css/CSSNamedImageValue.cpp:
* Source/WebCore/css/CSSNamedImageValue.h:
* Source/WebCore/css/CSSPaintImageValue.cpp:
* Source/WebCore/css/CSSPaintImageValue.h:
Move most of the logic and behaviors out of these classes and into
their Style counterparts. In some cases, member variable names were
expanded to make it more clear what they could hold in the surprising
cases where it was either an image or the `none` identifier.

* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::equals const):
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
Remove unused creation function and subsequent dead code removing
it creates.

* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
As this change focused on the generated image types, this class remains
relatively the same, but has its create methods cleaned up and replaces
the valueWithStylesResolved with the new direct createStyleImage. The
ResolvedURL type was also moved to CSSParserContext so it can be included
in more places with out cycles.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapNinePieceImage):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::isImageShape):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
Now that there isn&apos;t a CSSImageGeneratorValue type, we use the existing
isImageGeneratorValue() function that does the same thing as an is&lt;&gt;.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForNinePieceImage):
(WebCore::ComputedStyleExtractor::currentColorOrValidColor):
(WebCore::contentToCSSValue):
(WebCore::shapePropertyValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
(WebCore::ComputedStyleExtractor::currentColorOrValidColor const): Deleted.
* Source/WebCore/css/ComputedStyleExtractor.h:
With the StyleImage subclasses taking on more responsibility, a bit more context
is now needed when generating computed values for them. To resolve this, we now
pass the currently computing RenderStyle to them in their renamed computedStyleValue()
methods (previously named cssValue()).

* Source/WebCore/css/StyleColor.cpp:
* Source/WebCore/css/StyleColor.h:
(WebCore::serializationForRenderTreeAsText): Deleted.
Remove unused function.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
Move ResolvedURL here from CSSImageValue to allow use in more places
with including CSSImageValue. It also makes more sense here as this
is where they are created.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeGradientColorStops):
Remove now removed third argument.
(WebCore::CSSPropertyParserHelpers::consumeCrossFade):
(WebCore::CSSPropertyParserHelpers::consumeFilterImage):
Add some notes and FIXMEs about the implementation and rename a few variables
for clarity.

* Source/WebCore/html/CanvasBase.cpp:
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasObserver::isStyleCanvasImage const):
(WebCore::CanvasObserver::isCanvasObserverProxy const): Deleted.
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::~HTMLCanvasElement):
Update to reflect that now the StyleCanvasImage and not the CSSCanvasValue::CanvasObserverProxy
is the observer for canvas() images.

* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::collectPresentationalHintsForAttribute):
Pass the initiatorName via the create function rather than an
explicit getter, now that the create function supports that. This
allows the type to move toward immutability.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
(WebCore::HTMLInputElement::createInnerTextStyle):
Now that StyleGradientImage is more than just a wrapper around CSSGradientValue, we can
use it directly for the image mask, bypassing the need for a CSS value entirely.

* Source/WebCore/rendering/style/StyleCachedImage.cpp:
(WebCore::StyleCachedImage::create):
(WebCore::StyleCachedImage::copyOverridingScaleFactor):
* Source/WebCore/rendering/style/StyleCachedImage.h:
While mostly unchanged behavior wise, update to conform to some of the
idiomatic norms of the StyleImage types, adding a typed equals() function
(used by the CSSPropertyAnimations), and an explicit copyOverridingScaleFactor
create function used by StyleMultiImage to avoid that class knowing that
getting the internal CSSValue was possible (since it won&apos;t be in the future).

* Source/WebCore/rendering/style/StyleCanvasImage.cpp:
* Source/WebCore/rendering/style/StyleCanvasImage.h:
* Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
* Source/WebCore/rendering/style/StyleCrossfadeImage.h:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.h:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
* Source/WebCore/rendering/style/StyleFilterImage.h:
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/rendering/style/StyleImageSet.cpp:
* Source/WebCore/rendering/style/StyleImageSet.h:
* Source/WebCore/rendering/style/StyleNamedImage.cpp:
* Source/WebCore/rendering/style/StyleNamedImage.h:
* Source/WebCore/rendering/style/StylePaintImage.cpp:
* Source/WebCore/rendering/style/StylePaintImage.h:
Move logic and state from the CSSValue counterparts, aiming to change as little
as possible. This means that in some cases we are storing the data in a non
optimal way. For example, we store gradient parameters as the CSSPrimitiveValues
to keep the current (and incorrect) behavior for color serialization. This
allows this change to be behavior neutral, with the fixes coming on their own.

* Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
* Source/WebCore/rendering/style/StyleGeneratedImage.h:
(WebCore::StyleGeneratedImage:: const):
Move the logic from CSSImageGeneratorValue for generated image caching / eviction
and client tracking. Also adds additional virtual functions for the generated images,
didAddClient/didRemoveClient to allow them to react to clients. This is used by
StyleCanvasImage, for example, to notify the inspector.

* Source/WebCore/rendering/style/StyleImage.h:
Rename cssValue() function to computedStyleValue() and pass in the computing RenderStyle.
This is needed now that the StyleImages don&apos;t store pointers to their original CSSValues
anymore.

* Source/WebCore/rendering/style/StyleMultiImage.cpp:
* Source/WebCore/rendering/style/StyleMultiImage.h:
Move ImageWithScale here (it was previously defined in CSSImageSetValue)
and replace the CSSValue with a StyleImage. This means we now can track
the image set options as StyleImages, allowing the removal of creates
in StyleMultiImage::load (except the one when you have a StyleCachedImage
with a different scale factor).

* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::createStyleImage):
(WebCore::Style::BuilderState::resolveImageStyles): Deleted.
* Source/WebCore/style/StyleBuilderState.h:
Remove resolveImageStyles entirely, merging resolution into style
creation, and converting to taking const CSSValue references,
removing the possibility for the confusing interim resolved state
to return.

* Source/WebCore/svg/SVGCursorElement.cpp:
* Source/WebCore/svg/SVGCursorElement.h:
Updated to use StyleCursorImage as the client.

Canonical link: <a href="https://commits.webkit.org/255897@main">https://commits.webkit.org/255897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f2f26ba2584bf75f09c1ae9051665da91b2b55e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93951 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3140 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24516 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103584 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163921 "Hash 5f2f26ba for PR 5431 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3156 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31379 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86273 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2261 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80374 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29278 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84187 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72234 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37768 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17711 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35638 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18976 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/4069 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39511 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41544 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1920 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41450 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38206 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -381,17 +377,10 @@ static inline FilterOperations blendFunc(const FilterOperations& from, const Fil
 }
 
-static inline RefPtr<StyleImage> blendFilter(CachedImage* image, const FilterOperations& from, const FilterOperations& to, const CSSPropertyBlendingContext& context)
+static inline RefPtr<StyleImage> blendFilter(RefPtr<StyleImage> inputImage, const FilterOperations& from, const FilterOperations& to, const CSSPropertyBlendingContext& context)
```
Why `RefPtr` instead of `Ref`?

weinig replied:
At the moment, I can't prove that inputImage is non-null due to the parser currently allowing using `none` for images in some cases. This is something I want to improve if possible, since the specs don't actually allow this and it doesn't generally seem useful to be able to do.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -494,51 +477,55 @@ static inline RefPtr<StyleImage> blendFunc(StyleImage* from, StyleImage* to, con
     if (is<StyleGeneratedImage>(*from) && is<StyleGeneratedImage>(*to)) {
-        CSSImageGeneratorValue& fromGenerated = downcast<StyleGeneratedImage>(*from).imageValue();
-        CSSImageGeneratorValue& toGenerated = downcast<StyleGeneratedImage>(*to).imageValue();
+        auto& fromGenerated = downcast<StyleGeneratedImage>(*from);
```
Could use `dynamicDowncast` instead. @cdumez seems to prefer it. Either way seems good. Alternatively might be nice to slightly cut down on the total number of locals by not always putting the result into a named local. It’s kind of a long function so cutting down on boilerplate would be a plus.

weinig replied:
I can't figure out how to work `dynamicDowncast` in here too well due to needing to both downcast multiple items, but taking a look at the whole thing, I was able to remove one of the outer conditionals (the one that only checks `if (is<StyleGeneratedImage>(*from) && is<StyleGeneratedImage>(*to))`.

darinadler replied:
I guess we need to invent variadic `dynamicDowncast`.

darinadler commented on Source/WebCore/css/CSSImageSetValue.cpp:
```diff
@@ -137,9 +60,30 @@ String CSSImageSetValue::customCSSText() const
+        ASSERT(is<CSSPrimitiveValue>(scaleFactorValue));
+
+        float scaleFactor = downcast<CSSPrimitiveValue>(scaleFactorValue)->floatValue(CSSUnitType::CSS_DPPX);
+        images.append({ state.createStyleImage(*imageValue), scaleFactor });
```
`uncheckedAppend` is called for here

darinadler commented on Source/WebCore/css/CSSToStyleMap.cpp:
```diff
@@ -473,7 +472,7 @@ void CSSToStyleMap::mapNinePieceImage(CSSValue* value, NinePieceImage& image)
 
     for (auto& current : borderImage) {
-        if (is<CSSImageValue>(current) || is<CSSImageGeneratorValue>(current) || is<CSSImageSetValue>(current))
+        if (is<CSSImageValue>(current) || is<CSSImageSetValue>(current) || current->isImageGeneratorValue())
```
Should use `*current` in the first two if we are also dereferencing unconditionally in the third predicate. Also below. But you said you tried not to do too much of that kind of thing.

weinig replied:
I am going to add an isImage function for this. The annoying aspect is that we currently sometimes treat CSS cursors as images, and sometimes don't. For most cases, we shouldn't, like here, but it makes the naming odd. Will add a good comment.

darinadler commented on Source/WebCore/css/parser/CSSParserContext.h:
```diff
@@ -37,7 +37,29 @@ namespace WebCore {
+{
+    return !(a == b);
+}
+
```
One less blank line please.

darinadler commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -181,13 +181,13 @@ HashSet<Element*> CanvasBase::cssCanvasClients() const
-        for (auto& entry : clients) {
-            if (RefPtr<Element> element = entry.key->element())
-                cssCanvasClients.add(element.get());
+        auto clients = downcast<StyleCanvasImage>(observer).clients().values();
```
Can we do this without the local please?

darinadler commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -181,13 +181,13 @@ HashSet<Element*> CanvasBase::cssCanvasClients() const
+        auto clients = downcast<StyleCanvasImage>(observer).clients().values();
+        for (auto client : clients) {
+            if (auto element = client->element())
+                cssCanvasClients.add(element);
```
Seems like this should be `*element`, but that requires changing the called function.

darinadler commented on Source/WebCore/rendering/style/StyleCachedImage.h:
```diff
@@ -36,16 +36,18 @@ class Document;
 public:
-    static Ref<StyleCachedImage> create(CSSImageValue& cssValue, float scaleFactor = 1);
+    static Ref<StyleCachedImage> create(Ref<CSSImageValue>, float scaleFactor = 1);
+    static Ref<StyleCachedImage> copyOverridingScaleFactor(StyleCachedImage&, float scaleFactor);
```
Unclear whether `copy` or `clone` is the best verb for this operation.

darinadler commented on Source/WebCore/rendering/style/StyleCanvasImage.cpp:
```diff
@@ -28,73 +28,124 @@
+bool StyleCanvasImage::knownToBeOpaque(const RenderElement&) const
 {
-    return m_imageGeneratorValue->knownToBeOpaque(renderer);
+    // FIXME: When CanvasRenderingContext2DSettings.alpha is implemented, this can be improved to check for it.
```
Would be even better if the comment explained why. “To optimize performance for some common cases” or whatever the reason is.

darinadler commented on Source/WebCore/rendering/style/StyleCanvasImage.cpp:
```diff
@@ -28,73 +28,124 @@
+    if (!m_element) {
+        m_element = document.getCSSCanvasElement(m_name);
+        if (!m_element)
+            return nullptr;
```
Is it OK that this keeps calling `getCSSCanvasElement` repeatedly if it returns `nullptr`?

darinadler commented on Source/WebCore/rendering/style/StyleCanvasImage.h:
```diff
@@ -26,42 +26,57 @@
-    Ref<CSSCanvasValue> m_imageGeneratorValue;
+    HTMLCanvasElement* element(Document&) const;
+
+    // The name of the canvas.
```
Seems like this comment should influence the name of the data member and then the comment would be removed.

darinadler commented on Source/WebCore/rendering/style/StyleCanvasImage.h:
```diff
@@ -26,42 +26,57 @@
+
+    // The name of the canvas.
+    String m_name;
+    // The document supplies the element and owns it.
```
We use `WeakPtr` for this kind of thing in most cases today.

darinadler commented on Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
```diff
@@ -27,74 +27,163 @@
+    if (fromImageSize == toImageSize)
+        return fromImageSize;
+
+    float percentage = m_percentage;
```
Why `float` instead of `double` or `auto`? Do we have test coverage of edge cases where they might differ? Clearly the size ends up `float` but does converting earlier have any effect on the result?

darinadler commented on Source/WebCore/rendering/style/StyleCrossfadeImage.h:
```diff
@@ -26,39 +26,55 @@
 
-    Ref<CSSCrossfadeValue> m_imageGeneratorValue;
+    // FIXME: Rather than caching and tracking the input image via CachedImages, we should
+    // instead use a new, StyleImage specific notification, to allow correct tracking of
```
Should not have a comma before “to” here.

darinadler commented on Source/WebCore/rendering/style/StyleCursorImage.cpp:
```diff
@@ -26,51 +26,123 @@
+        static_cast<int>(std::round(cursorElement.y().value(lengthContext)))
+    };
+
+    // FIXME: Why doesn't this funtion check for a change to the href of the cursor element? Why would we dynamically track changes to x/y but not href?
```
Misspelled “function” in this comment.

darinadler commented on Source/WebCore/rendering/style/StyleCursorImage.h:
```diff
@@ -25,31 +25,44 @@
-struct ImageWithScale;
+class CSSValue;
+class Document;
+class WeakPtrImplWithEventTargetData;
```
Comes after “S” in alphabetical order.

darinadler commented on Source/WebCore/rendering/style/StyleFilterImage.cpp:
```diff
@@ -27,74 +27,154 @@
+        return true;
+
+    if (!m_image) {
+        if (other.m_image)
```
This if statement is always true. Can just return false unconditionally here and may want to use `else if` below. Or see my better idea.

darinadler commented on Source/WebCore/rendering/style/StyleFilterImage.cpp:
```diff
@@ -27,74 +27,154 @@
+            return false;
+    }
+    
+    return true;
```
I think I would probably just write:
```
    return m_image == other.m_image
        || m_image && other.m_image && *m_image == *other.m_image;
```
Might even want a named function or function template for this pattern.

darinadler commented on Source/WebCore/rendering/style/StyleFilterImage.cpp:
```diff
@@ -27,74 +27,154 @@
+        return &Image::nullImage();
+
+    if (size.isEmpty())
+        return nullptr;
```
Probably needs a comment to explain why this is `nullptr` and so many other cases are `nullImage`.

darinadler commented on Source/WebCore/rendering/style/StyleFilterImage.cpp:
```diff
@@ -27,74 +27,154 @@
-{
-    return m_imageGeneratorValue->clients().contains(&renderer);
+    for (auto& client : clients().values())
+        client->imageChanged(static_cast<WrappedImagePtr>(this));
```
Surprised we need an explicit cast (here and elsewhere with the same idiom). Also might want one a bit more type safe than calling `static_cast` directly.

darinadler commented on Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
```diff
@@ -28,13 +28,77 @@
-    
-StyleGeneratedImage::StyleGeneratedImage(Type type, bool fixedSize)
+
+static const Seconds timeToKeepCachedGeneratedImages { 3_s };
```
`constexpr`

darinadler commented on Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
```diff
@@ -63,10 +127,39 @@ FloatSize StyleGeneratedImage::imageSize(const RenderElement* renderer, float mu
+
+    m_clients.add(&renderer);
+
+    this->didAddClient(renderer);
```
Is explicit `this` needed?

darinadler commented on Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
```diff
@@ -63,10 +127,39 @@ FloatSize StyleGeneratedImage::imageSize(const RenderElement* renderer, float mu
+    if (!m_clients.remove(&renderer))
+        return;
+
+    this->didRemoveClient(renderer);
```
Is explicit `this` needed?

darinadler commented on Source/WebCore/rendering/style/StyleGeneratedImage.h:
```diff
@@ -23,31 +23,60 @@
 
 #pragma once
 
+#include "FloatSize.h"
```
I think `FloatSizeHash.h` pulls this in so we don’t need both.

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -27,74 +27,1101 @@
-    return m_imageGeneratorValue->knownToBeOpaque(renderer);
+    bool hasColorFilter = renderer.style().hasAppleColorFilter();
+    for (auto& stop : m_stops) {
+        Color color = stop.resolvedColor;
```
A shame to make a copy when the color is extended and `hasColorFilter` is false.

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -27,74 +27,1101 @@
+
+class ConicGradientAdapter {
+public:
+    float gradientLength() const { return 1; }
```
Maybe `static`, `constexpr`, or both?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -27,74 +27,1101 @@
+class ConicGradientAdapter {
+public:
+    float gradientLength() const { return 1; }
+    float maxExtent(float, float) const { return 1; }
```
Maybe `static`, `constexpr`, or both?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -27,74 +27,1101 @@
+        return gradientSize.diagonalLength();
+    }
+
+    float maxExtent(float, float) const { return 1; }
```
Maybe `static`, `constexpr`, or both?

darinadler commented on Source/WebCore/rendering/style/StyleGradientImage.h:
```diff
@@ -26,39 +26,92 @@
+    CSSGradientType m_gradientType;
+    CSSGradientColorInterpolationMethod m_colorInterpolationMethod;
+    Vector<Stop> m_stops;
+    mutable std::optional<bool> m_hasColorDerivedFromElement;
```
I thought you were a fervent believer in “no `optional<bool>`”?

weinig replied:
:( Just one of many things I had to convince myself not to cleanup in this pass (this is just moved from CSSGradientValue). This bool goes away entirely in the next change, so it wasn't worth rocking the boat in this one.

darinadler commented on Source/WebCore/rendering/style/StyleImageSet.cpp:
```diff
@@ -47,14 +50,57 @@ bool StyleImageSet::operator==(const StyleImage& other) const
+    ImageWithScale image;
+    size_t numberOfImages = m_images.size();
+    for (size_t i = 0; i < numberOfImages; ++i) {
+        image = m_images.at(i);
```
Range-based for loop instead?

darinadler commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1563,7 +1563,7 @@ inline void BuilderCustom::applyValueContent(BuilderState& builderState, CSSValu
     bool didSet = false;
     for (auto& item : downcast<CSSValueList>(value)) {
-        if (is<CSSImageGeneratorValue>(item) || is<CSSImageSetValue>(item) || is<CSSImageValue>(item)) {
+        if (is<CSSImageValue>(item) || is<CSSImageSetValue>(item) || item->isImageGeneratorValue()) {
```
Should this use the `isImageShape` function?

weinig replied:
Going to replace isImageShape and these checks with a single isImage() function (see above).

Commit:
Move logic for generated images from CSSValue types to the StyleImage types
https://bugs.webkit.org/show_bug.cgi?id=246594
rdar://101225901

Reviewed by Darin Adler.

In a step toward creating a comprehensive layering between the CSSValue types
and the Style types they resolve to, this change moves all of the logic that
used to be in the CSSValue derived types for generated images into their new
Style counterparts. This leaves the CSSValue side as immutable representations
of the parsed CSS (or an immutable representation of a computed CSS property)
that is capable of only a few operations:
   - equality
   - serialization
   - resolution to Style counterpart

This initial move does a pretty close approximation of just moving the logic
and state, in some cases having the Style type continue to store CSSValues,
but subsequent changes should change this so that we do conversion to lower
level types at creation time (e.g. use Length rather than CSSPrimitiveValue).
The one major exception is that we now reference other images as StyleImages
rather than CSSValues in the places where the types require input images
such as StyleCrossfadeImage and StyleFilterImage. This puts us in a strong
position to fix a bunch of long standing issues with animating between
various combinations of nested images.

Due to the simplification on the CSSValue side, we no longer need the
CSSImageGeneratorValue intermediate class, so it has been removed.

This change does not update CSSImageValue in the same way as the generated
image classes, as that has additional complications that should be handled
on their own.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSImageGeneratorValue.cpp: Removed.
* Source/WebCore/css/CSSImageGeneratorValue.h: Removed.
Remove CSSImageGeneratorValue.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFilter):
(WebCore::crossfadeBlend):
(WebCore::blendFunc):
Update to use StyleImages (and subclasses thereof) rather than the CSSValue
types, making the logic much closer to other blend primitives and allowing
the removal of a lot of unnecessary temporary allocations.

* Source/WebCore/css/CSSCanvasValue.cpp:
* Source/WebCore/css/CSSCanvasValue.h:
* Source/WebCore/css/CSSCrossfadeValue.cpp:
* Source/WebCore/css/CSSCrossfadeValue.h:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFilterImageValue.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/CSSImageSetValue.cpp:
* Source/WebCore/css/CSSImageSetValue.h:
* Source/WebCore/css/CSSNamedImageValue.cpp:
* Source/WebCore/css/CSSNamedImageValue.h:
* Source/WebCore/css/CSSPaintImageValue.cpp:
* Source/WebCore/css/CSSPaintImageValue.h:
Move most of the logic and behaviors out of these classes and into
their Style counterparts. In some cases, member variable names were
expanded to make it more clear what they could hold in the surprising
cases where it was either an image or the `none` identifier.

* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::equals const):
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
Remove unused creation function and subsequent dead code removing
it creates.

* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
As this change focused on the generated image types, this class remains
relatively the same, but has its create methods cleaned up and replaces
the valueWithStylesResolved with the new direct createStyleImage. The
ResolvedURL type was also moved to CSSParserContext so it can be included
in more places with out cycles.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapNinePieceImage):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::isImageShape):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
Now that there isn't a CSSImageGeneratorValue type, we use the existing
isImageGeneratorValue() function that does the same thing as an is<>.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForNinePieceImage):
(WebCore::ComputedStyleExtractor::currentColorOrValidColor):
(WebCore::contentToCSSValue):
(WebCore::shapePropertyValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
(WebCore::ComputedStyleExtractor::currentColorOrValidColor const): Deleted.
* Source/WebCore/css/ComputedStyleExtractor.h:
With the StyleImage subclasses taking on more responsibility, a bit more context
is now needed when generating computed values for them. To resolve this, we now
pass the currently computing RenderStyle to them in their renamed computedStyleValue()
methods (previously named cssValue()).

* Source/WebCore/css/StyleColor.cpp:
* Source/WebCore/css/StyleColor.h:
(WebCore::serializationForRenderTreeAsText): Deleted.
Remove unused function.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
Move ResolvedURL here from CSSImageValue to allow use in more places
with including CSSImageValue. It also makes more sense here as this
is where they are created.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeGradientColorStops):
Remove now removed third argument.
(WebCore::CSSPropertyParserHelpers::consumeCrossFade):
(WebCore::CSSPropertyParserHelpers::consumeFilterImage):
Add some notes and FIXMEs about the implementation and rename a few variables
for clarity.

* Source/WebCore/html/CanvasBase.cpp:
* Source/WebCore/html/CanvasBase.h:
(WebCore::CanvasObserver::isStyleCanvasImage const):
(WebCore::CanvasObserver::isCanvasObserverProxy const): Deleted.
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::~HTMLCanvasElement):
Update to reflect that now the StyleCanvasImage and not the CSSCanvasValue::CanvasObserverProxy
is the observer for canvas() images.

* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::collectPresentationalHintsForAttribute):
Pass the initiatorName via the create function rather than an
explicit getter, now that the create function supports that. This
allows the type to move toward immutability.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
(WebCore::HTMLInputElement::createInnerTextStyle):
Now that StyleGradientImage is more than just a wrapper around CSSGradientValue, we can
use it directly for the image mask, bypassing the need for a CSS value entirely.

* Source/WebCore/rendering/style/StyleCachedImage.cpp:
(WebCore::StyleCachedImage::create):
(WebCore::StyleCachedImage::copyOverridingScaleFactor):
* Source/WebCore/rendering/style/StyleCachedImage.h:
While mostly unchanged behavior wise, update to conform to some of the
idiomatic norms of the StyleImage types, adding a typed equals() function
(used by the CSSPropertyAnimations), and an explicit copyOverridingScaleFactor
create function used by StyleMultiImage to avoid that class knowing that
getting the internal CSSValue was possible (since it won't be in the future).

* Source/WebCore/rendering/style/StyleCanvasImage.cpp:
* Source/WebCore/rendering/style/StyleCanvasImage.h:
* Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
* Source/WebCore/rendering/style/StyleCrossfadeImage.h:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.h:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
* Source/WebCore/rendering/style/StyleFilterImage.h:
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/rendering/style/StyleImageSet.cpp:
* Source/WebCore/rendering/style/StyleImageSet.h:
* Source/WebCore/rendering/style/StyleNamedImage.cpp:
* Source/WebCore/rendering/style/StyleNamedImage.h:
* Source/WebCore/rendering/style/StylePaintImage.cpp:
* Source/WebCore/rendering/style/StylePaintImage.h:
Move logic and state from the CSSValue counterparts, aiming to change as little
as possible. This means that in some cases we are storing the data in a non
optimal way. For example, we store gradient parameters as the CSSPrimitiveValues
to keep the current (and incorrect) behavior for color serialization. This
allows this change to be behavior neutral, with the fixes coming on their own.

* Source/WebCore/rendering/style/StyleGeneratedImage.cpp:
* Source/WebCore/rendering/style/StyleGeneratedImage.h:
(WebCore::StyleGeneratedImage:: const):
Move the logic from CSSImageGeneratorValue for generated image caching / eviction
and client tracking. Also adds additional virtual functions for the generated images,
didAddClient/didRemoveClient to allow them to react to clients. This is used by
StyleCanvasImage, for example, to notify the inspector.

* Source/WebCore/rendering/style/StyleImage.h:
Rename cssValue() function to computedStyleValue() and pass in the computing RenderStyle.
This is needed now that the StyleImages don't store pointers to their original CSSValues
anymore.

* Source/WebCore/rendering/style/StyleMultiImage.cpp:
* Source/WebCore/rendering/style/StyleMultiImage.h:
Move ImageWithScale here (it was previously defined in CSSImageSetValue)
and replace the CSSValue with a StyleImage. This means we now can track
the image set options as StyleImages, allowing the removal of creates
in StyleMultiImage::load (except the one when you have a StyleCachedImage
with a different scale factor).

* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::createStyleImage):
(WebCore::Style::BuilderState::resolveImageStyles): Deleted.
* Source/WebCore/style/StyleBuilderState.h:
Remove resolveImageStyles entirely, merging resolution into style
creation, and converting to taking const CSSValue references,
removing the possibility for the confusing interim resolved state
to return.

* Source/WebCore/svg/SVGCursorElement.cpp:
* Source/WebCore/svg/SVGCursorElement.h:
Updated to use StyleCursorImage as the client.

Canonical link: https://commits.webkit.org/255897@main
